### ISSUE TITLE
prefers-color-scheme

### ISSUE THREAD

**Issue by kaihendry** [kognise/water.css#25]
https://developer.apple.com/documentation/safari_release_notes/safari_12_1_release_notes

Be good if you can support Safari media queries so that if a user wants a dark scheme, they would get one. And if a user doesn't, they would get a light theme instead?

**Comment by kognise** [comment]
I keep getting a billion requests to add this support. Time to think about possible ways to implement it!

**Comment by kylejrp** [comment]
This worked for me in Firefox 67. It defaults to light, unless your browser supports prefers-color-scheme and your preference is also dark:

```
<link id="stylesheet" rel="stylesheet" href="dist/light.css">
<link id="stylesheet" rel="stylesheet" href="dist/dark.css" media="(prefers-color-scheme: dark)">
```

Alternatively, for a dark default unless your preference is light:
```
<link id="stylesheet" rel="stylesheet" href="dist/dark.css">
<link id="stylesheet" rel="stylesheet" href="dist/light.css" media="(prefers-color-scheme: light)">
```

We could also do this in Sass, but it would create an extra file that's twice the size. Eg:
```
/* auto-theme.scss */
@import './light'

@media (prefers-color-scheme: dark) {
    @import './dark'
}
```

Doing it inside the `<link  />` tag is better because (I think) it saves a network request when we don't need the other theme.

**Comment by kaihendry** [comment]
This tact appears to work. Just experimenting with it on my homepage: https://hendry.iki.fi/

It works on Safari on a MBP. Now I am wondering how to tell my IOS phone my preference is dark mode, along with my Archlinux machine's chrome.

**Comment by kylejrp** [comment]
For fun, I tried combining these both into a new file called `water.sass`:
```
@import "./light";

@media (prefers-color-scheme: dark) {
    @import "./dark";
}
```

While the main size is fairly large, it's really small when gzipped:
` Original size: 6582 bytes Result size: 1576 bytes`

@kognise Do you think this would be a valuable file to have? We could just have people use `water.css` or `water.min.css` and then there's no thoughts to picking a theme. The downside is our main advertised file is fairly large.

If we don't want to advertise a huge file, we could reword this to be `auto.css` and advertise it as an auto-theme detection css file, but with the disadvantage of a larger file size.

**Comment by kognise** [comment]
@kylejrp At least in my opinion, I would just prefer recommending using the `media` attribute and two link elements. If we *have* to have a file for this I would prefer the `auto.css` approach.

What would you recommend?

**Comment by kylejrp** [comment]
Again I think this comes down to valuing size vs valuing performance. Despite having a combined file that's larger, it's actually more performant than one with link elements.

Here's the timeline for a combined file:
1. Download `water.css` (~30ms TTFB, ~5ms download)
2. Show page

Here's the timeline for a split out file:
1. Download `auto.css` that uses link elements (30ms TTFB, 1ms download)
2. Calculate the rules
3. Download `dark.css` (30ms TTFB, 1ms download) and if prefers-color-scheme is light, also download (in parallel?)`light.css` (30ms TTFB, 1ms download)

Even if we assume that `dark.css`/`light.css` can be downloaded in parallel, `auto.css` with links would take ~60+ms to show any content, compared to a combined `water.css` which shows content after ~35ms.

🤷‍♂️

**Comment by kognise** [comment]
@kylejrp I'm sorry, I'm not sure I'm fully understanding - what are steps 1 and two for the split out file?

**Comment by kylejrp** [comment]
Here's my explanation with --verbose turned on!

1. [0ms] Browser downloads `auto.css` that uses link elements. This takes 30ms to start getting a response from the CDN. The actual download takes ~5ms. The file looks something like this:
```
@import url("./dark.css");
@import url("./light.css") (prefers-color-scheme: light);
```
2. [35ms] Browser processes `auto.css`. It realizes it needs to import `dark.css` and `light.css`. It doesn't have these downloaded yet, so it makes two requests.
3. [35ms] Let's assume that it makes two network requests in parallel. It takes 30ms to start getting a response from the CDN for both files, and ~5ms to download them.
4. [70ms] Browser processes `dark.css` and `light.css`. It starts showing the styles in the browser.

The key part is that the browser doesn't know about `dark.css` and `light.css` until it finishes getting the response for `auto.css`.

**Comment by kylejrp** [comment]
Although I guess just using this solves the "waiting on the original request" problem:
```
<link id="stylesheet" rel="stylesheet" href="dist/light.css">
<link id="stylesheet" rel="stylesheet" href="dist/dark.css" media="(prefers-color-scheme: dark)">
```

**Comment by kognise** [comment]
Ahhhh. So then 2 link elements is probably the way to go?

**Comment by kylejrp** [comment]
Yeah, I guess so! I was hoping for an all-in-one solution so it would be easy to communicate, but now we have a bunch of options to communicate in the docs:

1. Dark theme only
2. Light theme only
3. Dark theme by default, light theme when prefers-color-scheme: light
4. Light theme by default, dark theme when prefers-color-scheme: dark

I'm not sure what the best way to communicate these choices is.

**Comment by kognise** [comment]
Agreed, there's a communication challenge now. Already I think it's pretty hard to find that there's a light theme on first glance in the docs. I don't have much time now; I'll send some ideas later today.

**Comment by jonaskuske** [comment]
We could also use "auto" CSS files but recommend adding `link[rel=preload]` when necessary.

So we'd have

- `dist/dark.css` (dark by default, light if `prefers-color-scheme: light`)
- `dist/light.css` (light by default, dark if `prefers-color-scheme: dark`)
- `dist/light-standalone.css` (only the light theme)
- `dist/dark-standalone.css` (only the dark theme)

All of them could be used with `link[rel=stylesheet]` on their own, but for `light.css` and `dark.css` we'd also recommend to add preload tags like:

```html
<link rel="preload" as="style" href=".../dark-standalone.css">
<link rel="preload" as="style" href=".../light-standalone.css" media="(prefers-color-scheme: light)">
```

Also, as my example above shows, I'd very much like it if the default CSS files would be the ones that respect user choice (`prefers-color-scheme`) and the versions that force a theme are the ones requiring extra (typing) effort by the devs, as in: have the more complicated names. 😀

**Comment by jonaskuske** [comment]
Another thing:  

If we also had "modern" versions of the CSS files, devs not targeting Internet Explorer and the likes wouldn't have to deal with these "loading order" problems at all.

We could just use normal CSS variables instead of Sass variables then, and the "auto" builds could ship both sets of variables (negligible additonal filesize) and toggle the variables based on the media query. 😅

**Comment by jonaskuske** [comment]
**Example interface for the different versions** 👀
https://codepen.io/21stCenturyJonas/full/JqjPgL
(assumes a modern build with CSS vars exists and uses npm links that'll come with #64 as they're shorter)

**Comment by kylejrp** [comment]
Oh wow, duh, CSS variables make so much sense for this. What a great idea. (We live in the future!)

We should pursue using CSS variables and supporting prefers-color-scheme. We can support IE11 and the likes later with the advanced options discussed in #27. 

The trickiest part here would be modifying our existing Sass/Gulp to handle this. Ideally, we just have Sass variables or CSS variables, not a mix of both. Here's a few options:
1. Switch all code to use CSS variables, use [https://www.npmjs.com/package/postcss-css-variables](postcss-css-variables) to "compile" these when necessary for browser support
2. Switch all code to use CSS variables, offer the [CSS variables polyfill](https://github.com/jhildenbiddle/css-vars-ponyfill) for browser support 
3. Keep Sass variables, compile these to CSS variables somehow (?)

I think 1 makes the most sense here, but there's some work to get this set up. I can try making a pull request that uses the CSS variables while still supporting legacy browsers where necessary.

This will be dependent on having nicer installation UI from #27 though.

**Comment by kylejrp** [comment]
I'm seeing Chrome 76-77 having support for this now:
https://caniuse.com/#feat=prefers-color-scheme

**Comment by jonaskuske** [comment]
This can be closed once the current master is published as a release. (version 2.0.0 I guess)

**Comment by kylejrp** [comment]
I'm closing this as the changes have been merged into master and we're just waiting on the release of 2.0.0.
